funsor/distribution.py

```diff
@@ -109,17 +110,16 @@ def _get_raw_dist(self):
         """
         Internal method for working with underlying distribution attributes
         """
-        if isinstance(self.value, Variable):
-            value_name = self.value.name
-        else:
-            raise NotImplementedError("cannot get raw dist for {}".format(self))
+        value_name = [name for name, domain in self.value.inputs.items()  # TODO is this right?
+                      if domain == self.value.output][0]
```

Comment:
This looks weird. Can you explain what's going on? When is `value_name != "value"`? The assertion in `self.__init__()` suggests `value_name == "value"` IIUC.

Reply:
`value_name` can potentially be anything, since `self.value` is a lazy expression. This logic is meant to solve the problem of identifying the value name when `self.value` is not a `Variable` or even has more than one input (which happens when constructing the Funsor version of a `TransformedDistribution`).

The simplest nontrivial example of the latter case would be an affine or power transform where the parameters are `funsor.Tensor`s with nontrivial `.inputs`, although these are not handled in this PR since `funsor.delta.solve` does not yet support inverting such expressions.

The main use for `value_name` in this PR is in `Distribution.unscaled_sample`, which needs to know `value_name` to construct a sample `Delta` with the correct `.inputs`.